client/core/trade.go

```diff
@@ -527,6 +527,50 @@ func (t *trackedTrade) counterPartyConfirms(match *matchTracker) (have, needed u
 	return
 }
 
+// deleteStaleCancelOrder checks if this trade has an associated cancel order,
+// and deletes the cancel order if it (i.e. the cancel order) stays at Epoch
+// status for 2 or more epochs after the close of the cancel order's epoch.
+// Deleting the stale cancel order from this trade makes it possible for the
+// client to re-attempt cancelling the order.
+//
+// NOTE:
+// Stale cancel orders would be Executed if their preimage was sent or Revoked
+// if their preimages was not sent. We cannot currently tell whether the cancel
+// order's preimage was revealed, so assume that the cancel order is Executed
+// but unmatched. Consider adding a order.PreimageRevealed field to ensure that
+// the correct final status is set for the cancel order.
+//
+// This method MUST be called with the trackedTrade mutex lock held for writes.
+func (t *trackedTrade) deleteStaleCancelOrder() {
+	if t.cancel == nil || t.metaData.Status != order.OrderStatusBooked {
+		log.Debugf("Ignoring delete cancel order attempt for order %v, status %v, pending cancel = %v",
+			t.ID(), t.metaData.Status, t.cancel != nil)
+		return
+	}
+
+	stamp := t.cancel.ServerTime
+	epoch := order.EpochID{Idx: encode.UnixMilliU(stamp) / t.epochLen, Dur: t.epochLen}
+	epochEnd := epoch.End()
+	if time.Since(epochEnd).Milliseconds() < int64(2*t.epochLen) {
+		return // not stuck, yet
```

Comment:
should `deleteStaleCancelOrder` be called during tick then?

Reply:
Perhaps, but an important prerequisite for the `deleteStaleCancelOrder` method is that the target order must be in status Booked. We're definitely sure of the status of orders after processing the connect response; but we could be wrong at other times. Specifically, if the cancel order's preimage/match/nomatch request was missed while connected to the DEX, it's possible that the target order's status has changed unknown to the client; and we might be wrong to delete the cancel order.

The way I see this working reliably from tick, is add a `hasStaleCancelOrder` method, collect the order ids of all trackedTrades that have stale cancel orders, request their current statuses from the DEX and only delete the stale cancel orders if the order_status response indicates that the target orders are still booked.

Having this done only on connect is still fine to me as it is a recovery measure, which I think should be better left for the dex authentication/connection procedure.

Anyways, the thought of being able to perform order status reconciliation without waiting for a DEX reconnect had come to mind. It'd be a case for #685, as that already lays the foundation for order status reconciliation during tick. Just need to plug in the order_status route and **core** will be quicker to notice stuck/stale orders and request status updates.

Reply:
Ok, we can let this reconciliation be for reconnect for now.